### PR TITLE
add README documentation

### PR DESCRIPTION
**Pull request by jackshirazi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,77 @@
...
+This project has two sub-projects
+
+* application - a standalone runnable example application that is here purely to provide an example target to instrument  
+* plugin - the plugin that instruments the application
+
+## Application sub-project
+
+The application consists of a webserver and a client that executes some requests against the webserver. A [webserver interface](blob/main/application/src/main/java/co/elastic/apm/example/webserver/ExampleHttpServer.java) and two implementations are provided:
+
+* [ExampleAlreadyInstrumentedHttpServer](blob/main/application/src/main/java/co/elastic/apm/example/webserver/ExampleAlreadyInstrumentedHttpServer.java) uses the com.sun.net.httpserver.HttpServer that is a standard part of the JDK to implement the webserver interface; the Elastic APM Java Agent already automatically instruments this technology, so this implementation is provided as a reference for checking logging and output
```
Comment (SylvainJuge) [inline README.md]:
```suggestion
* [ExampleAlreadyInstrumentedHttpServer](blob/main/application/src/main/java/co/elastic/apm/example/webserver/ExampleAlreadyInstrumentedHttpServer.java) uses the `com.sun.net.httpserver.HttpServer` that is a standard part of the JDK to implement the webserver interface; the Elastic APM Java Agent already automatically instruments this technology, so this implementation is provided as a reference for checking logging and output
```

I don't remember, is this one available in all versions of the JDK ?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,77 @@
...
+
+Note the application implementation is deliberately simple (eg System.out instead of a logging framework) to keep it as easy to understand as possible.
+
+## Plugin sub-project
+
+The plugin consists of a [single file](blob/main/plugin/src/main/java/co/elastic/apm/example/webserver/plugin/ExampleHttpServerInstrumentation.java) holding the custom instrumentation, several classes for regression testing, and a pom that builds the correct plugin jar. The details of the plugin project are explained in the articles [creating the instrumentation](https://www.elastic.co/blog/create-your-own-instrumentation-with-the-java-agent-plugin) and [regression testing it](https://www.elastic.co/blog/create-your-own-instrumentation-with-the-java-agent-plugin).
+
+## Building
+
+The full project can be built by cloning to your local system, changing to the root directory, and running `mvn clean install`.
```
Comment (SylvainJuge) [inline README.md:36]: [minor] Maybe add the prerequisites: git, maven and JDK installed.

Also providing the build instructions as as shell script block that allows any lazy person like me to copy-paste:
```
git clone ....
cd ...
mvn clean install
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,77 @@
...
+
+Each sub-project can also be separately built the same way (changing to the sub-project root directory and running `mvn clean install`).
+
+## Running
+
+You need an Elastic APM Java Agent jar (the latest version is recommended, but at least version 1.31.0). Additionally an Elastic APM server is recommended, though not required (communications to the server will be dropped if it's unavailable).
```
Comment (SylvainJuge) [inline README.md:50]: [minor] add link (or curl command) to download the latest agent.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,77 @@
...
+
+Assuming you have set the jar location in `AGENT_JAR` and if available the APM server is specified in `APM_SERVER_URL` and `APM_SECRET_TOKEN`, you can run the application with agent *but no plugin*, from the project root directory, as follows
+
+```aidl
+export AGENT_JAR=...
+export APM_SERVER_URL=...
+export APM_SECRET_TOKEN=...
+
+java -Delastic.apm.server_url=$APM_SERVER_URL -Delastic.apm.secret_token=$APM_SECRET_TOKEN -javaagent:$AGENT_JAR -Delastic.apm.log_level=DEBUG -Delastic.apm.service_name=ExampleClient-ExampleAlreadyInstrumentedHttpServer -cp application/target/application-*.jar co.elastic.apm.example.webserver.ExampleClient
+java -Delastic.apm.server_url=$APM_SERVER_URL -Delastic.apm.secret_token=$APM_SECRET_TOKEN -javaagent:$AGENT_JAR -Delastic.apm.log_level=DEBUG -Delastic.apm.service_name=ExampleClient-ExampleBasicHttpServer -cp application/target/application-*.jar co.elastic.apm.example.webserver.ExampleClient
```
Comment (SylvainJuge) [inline README.md:79]: [minor] We could probably make this a bit more straightforward for the end user by providing a ready-to-use shell script as it's easy to make a typo or copy-paste errors with long command lines like this.

For users that are on Windows asking them to run it with WSL/bash should not be a big ask as they are developpers.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,77 @@
 # apm-agent-java-plugin-example
 Example of instrumentation which plugs in to the Elastic Java Agent and instruments an application
+
+## Overview
+
+The [Elastic APM Java Agent](https://github.com/elastic/apm-agent-java/) is a Java agent that automatically measures the performance of your application and tracks errors. [Full documentation is available](https://www.elastic.co/guide/en/apm/agent/java/current/intro.html).
+
+This project provides a detailed example of using the Elastic APM Java Plugin API to add custom instrumentation in to the agent, which the agent will automatically apply.
```
Comment (AlexanderWert) [inline README.md]:
```suggestion
This project provides a detailed example of using the Elastic APM Java Plugin API to add custom instrumentation in to the agent, which the agent will automatically apply without the need for users to touch the target application's code.
```

File: README.md (hunk 4 above)
Comment (AlexanderWert) [reply README.md:79]: +1

File: README.md (hunk 1 above)
Comment (jackshirazi) [reply README.md]: no, all versions of hotspot and most versions of the others

File: README.md (hunk 2 above)
Comment (jackshirazi) [reply README.md:36]: thanks! added

File: README.md (hunk 3 above)
Comment (jackshirazi) [reply README.md:50]: thanks! added

File: README.md (hunk 4 above)
Comment (jackshirazi) [reply README.md:79]: thanks! added

**Pull request merged**